packages/react-bindings/src/styles/resolveStylesAndClasses.tsx

```diff
@@ -3,71 +3,158 @@ import {
   isDebugEnabled,
   ICSSInJSStyle,
   ComponentStyleFunctionParam,
+  ComponentSlotStylesResolved,
+  ThemePrepared,
 } from '@fluentui/styles'
 import { ComponentSlotClasses } from '../styles/types'
 
-// Both resolvedStyles and classes are objects of getters with lazy evaluation
+export type ResolveStylesResult = {
+  resolvedStyles: ComponentSlotStylesResolved
+  resolvedStylesDebug: Record<string, { styles: Object }[]>
+  classes: ComponentSlotClasses
+}
+
+// TODO: Describe me
+
+const classesCache = new WeakMap<ThemePrepared, Record<string, string>>()
+const stylesCache = new WeakMap<ThemePrepared, Record<string, ICSSInJSStyle>>()
+
+/**
+ * Both resolvedStyles and classes are objects of getters with lazy evaluation
+ *
+ * Additionally if the cacheEnabled option is provided, than the resolved styles
+ * and classes are caching the results in WeakMaps. The key of the maps contains the following:
+ * - theme
+ * - displayName
+ * - slot name
+ * - styling props
+ * - rtl mode
+ * - disable animations mode
+ */
 const resolveStylesAndClasses = (
   mergedStyles: ComponentSlotStylesPrepared,
   styleParam: ComponentStyleFunctionParam,
   renderStyles: (styles: ICSSInJSStyle) => string,
-): {
-  resolvedStyles: ICSSInJSStyle
-  resolvedStylesDebug: Record<string, { styles: Object }[]>
-  classes: ComponentSlotClasses
-} => {
+  cacheEnabled?: boolean | undefined,
+  displayName?: string,
+  theme?: ThemePrepared,
+  props?: object,
+): ResolveStylesResult => {
   const resolvedStyles: Record<string, ICSSInJSStyle> = {}
   const resolvedStylesDebug: Record<string, { styles: Object }[]> = {}
   const classes: Record<string, string> = {}
 
+  if (cacheEnabled && theme) {
+    if (!stylesCache.has(theme)) {
+      stylesCache.set(theme, {})
+    }
+    if (!classesCache.has(theme)) {
+      classesCache.set(theme, {})
+    }
+  }
+
   Object.keys(mergedStyles).forEach(slotName => {
     // resolve/render slot styles once and cache
-    const cacheKey = `${slotName}__return`
+    const lazyEvaluationKey = `${slotName}__return`
+    const stylesCacheKey =
+      cacheEnabled && displayName && props
+        ? `${displayName}: ${slotName}: ${JSON.stringify(props)}${styleParam.rtl}${
```

Comment:
`JSON.stringify` will not work as expected as the output depends on key insertion order.

```js
JSON.stringify({a: 1, b:2}) !== JSON.stringify({b: 2, a:1})
```

Reply:
We are always sending the props in the same order in useStyles, that's why we can use JSON.stringify. See some example, for the checkbox component:
```
  const { classes, styles: resolvedStyles } = useStyles(Checkbox.displayName, {
    className: Checkbox.className,
    mapPropsToStyles: () => ({
      checked: state.checked,
      disabled,
      labelPosition,
      toggle,
    }),
    mapPropsToInlineStyles: () => ({
      className,
      design,
      styles,
      variables,
    }),
    rtl: context.rtl,
  })
```
the order will always be 
```
      checked,
      disabled,
      labelPosition,
      toggle,
```

Reply:
🤦‍♂ my bad